### ISSUE TITLE
Should a generative function in the built-in modelling language be allowed to introspect a limited subset of its own choice map?

### ISSUE THREAD

**Issue by bzinberg** [probcomp/Gen.jl#159]
# Background
Generative functions written in the built-in modelling language have a body, which may make calls to other generative functions.  There is currently no expression that can be written within this body to "get my own choice map, as produced by the generative code in the body up to this point." This is intentional, as allowing that introspection would introduce a lot of complexity and might break the conceptual coherency of Gen.

# But what about
However, my understanding is that this complexity comes only from the case where a function tries to introspect the sub-choicemaps produced by calls to other generative functions (which may or may not be written in the built-in modelling language).  As far as I can see, the conceptual issues do _not_ come up if we instead only want the ability to "get a subset of my own choice map, consisting of what has been produced in the body up to this point, excluding the submaps produced by calls to other generative functions."

The reason I think it would be nice to have such a construct is exemplified by the following pseudocode for a joint generative model of scenes and images:

```julia
@gen function generateSceneAndImage()
  width = @trace(uniform(0, 1), :width)
  height = @trace(uniform(0, 1), :height)
  @trace(stochasticRenderRectangle(width, height), :rendering)
end
```

In the above code, we need to both tag each parameter for inclusion in the trace, and store it in a local variable for the call to `stochasticRenderRectangle`.  This is code and concept duplication (see https://github.com/probcomp/Cora/issues/23).  If we had the above construct, we could instead write simply

```julia
@gen function generateSceneAndImage()
  @trace(uniform(0, 1), :width)
  @trace(uniform(0, 1), :height)
  @trace(stochasticRenderRectangle(get_choices_so_far_shallow()), :rendering)
end
```

The comparative simplicity of the second version becomes more pronounced when the number of parameters grows larger.

**Comment by alex-lew** [comment]
Just discussed with Ben.

We agreed that this problem would be addressed in a more straightforward manner by a planned back-port of MetaGen's syntax to Gen, namely:

```
CURRENT SYNTAX: @trace(random_expr, addr_expr)
PROPOSED SYNTAX: {addr_expr} ~ random_expr
PROPOSED SUGAR: 
x ~ random_expr
desugars to
x = {:x} ~ random_expr
```

Then the example would become

```julia
@gen function generateSceneAndImage()
  width ~ uniform(0, 1)
  height ~ uniform(0, 1)
  rendering ~ stochasticRenderRectangle(width, height)
end
```

We also discussed that the semantics of "read my own choicemap" during `generate` or `update` might require some thought.

**Comment by bzinberg** [comment]
Just remembered another issue: What if I want to use the same function `stochasticRenderRectangle` both inside and outside the generative function body, where in the latter case the data I have on hand will be a trace of the scene model?

**Comment by alex-lew** [comment]
Hmm. I think my recommendation would be to do what you would do in ordinary programming: have `generateRectangle` return a `Rectangle` struct. This can then be passed as the input to `stochasticRenderRectangle`.

```julia
@gen function generateScene()
  width ~ uniform(0, 1)
  height ~ uniform(0, 1)
  Rectangle(width, height)
end

@gen function generateAndRenderScene()
  scene ~ generateScene()
  rendering ~ stochasticRenderRectangle(scene)
end

t = simulate(generateAndRenderScene, ())
stochasticRenderRectangle(t[:scene])
```

I think it's inadvisable to think of the choicemap as necessarily existing while the model is running: for example, when you just run `f(x)`, no choicemap for `f` is generated (which is why it's faster than calling `simulate`). I also think it's not a great idea to be passing choicemaps around in model code unless the model code is doing inference (a la MetaGen). Choicemaps are inference-time objects, and I like the idea of thinking of your models as being interpretable as ordinary Julia programs when you're not doing inference about them. Choicemaps are not as efficient as structs or named tuples or even dictionaries with accurate type information--sometimes the slowdown is quite large (3x between typed dict and choicemap when I was last experimenting with this);  when possible, I think it makes sense to use better data structures.

As an aside, in MetaGen, because inference algorithms _are_ models, you can write `t ~ traced(f(x))`, which calls `f` but returns its trace, instead of its return value. This is useful when using the trace as input to an inference algorithm that expects traces. I'm not convinced it should serve as a replacement for passing around local variables in general, though.

**Comment by bzinberg** [comment]
Thanks Alex for this explanation. 

> Hmm. I think my recommendation would be to do what you would do in ordinary programming: have generateRectangle return a Rectangle struct.

The `Rectangle` struct makes sense for the current example, and formally, makes sense in all the cases I was considering.  In more elaborate versions, however, I don't think there will always be an object-oriented interpretation of the struct (which is ok).  The struct would instead represent "the arguments to this specific function" (or other times, "the parameters of this specific model"), and would effectively replace the type signature of the function.  In C++, Java, etc., this is (i.m.e.) a pattern that can be employed to varying degrees in different use cases as a judgment call.

Your above suggestion, if I'm understanding correctly, is to always use the struct version, because then the argument/parameters struct can be used both within the function body and (once https://github.com/probcomp/Gen/pull/153 is merged) at inference time, via the new aux data capability.  The benefit of that style is to eliminate duplication of the modelling logic within inference code, while still allowing type-aware compilation which produces much more efficient low-level code.  Does that sound about right?

**Comment by alex-lew** [comment]
> Your above suggestion, if I'm understanding correctly, is to always use the struct version

Or, if a struct feels like overkill, a tuple would also do :-) You can then keep the function as is and use splatting to pass the arguments in.

**Comment by bzinberg** [comment]
True.  Would often have to specify the tuple type explicitly, to avoid the situation where an overly specific type is inferred and this breaks Gen's assumption that the type of data at a given address doesn't change?  (Though maybe that assumption is not as necessary for aux data, idk) (Edit: actually maybe the type is chosen at compile time and the same one is used for every call to the GF, idk)
